### DISCUSSION TITLE
WPF DataGrid memory leak in VS2022

### DISCUSSION THREAD

**Discussion by vsfeedback** [dotnet/wpf#5785] [Q&A]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/WPF-DataGrid-memory-leak-in-VS2022/1559338)._

---
[severity:I'm unable to use this version] [regression] [worked-in:16.11.3]
[Original Source](https://stackoverflow.com/questions/69643694/wpf-datagrid-memory-leak/69645193)
I have this code, and no matter what I do I keep getting a memory leak from the DataGrid data update. I've been looking at all the other answers to this problem here for days and I haven't found anything that works for me. I have a WPF window and this code to update the data (it is a small version of what happens in the real code, but I get the same bug). Works fine in VS2019.

I have this WPF code:

```
<Window x:Class="TremendoMemoryLeak.MainWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:local="clr-namespace:TremendoMemoryLeak"
        mc:Ignorable="d"
        DataContext="{Binding RelativeSource={RelativeSource Self}}"
        Title="MainWindow" Height="450" Width="800">
    <Grid>
        <DataGrid ItemsSource="{Binding Hmis, Mode=OneWay}" AutoGenerateColumns="False" IsReadOnly="True">
            <DataGrid.Columns>
                <DataGridTemplateColumn Header="Estado">
                    <DataGridTemplateColumn.CellTemplate>
                        <DataTemplate>
                            <Ellipse Fill="{Binding Estado}" HorizontalAlignment="Left" Height="25" Stroke="Black" VerticalAlignment="Top" Width="25"/>
                        </DataTemplate>
                    </DataGridTemplateColumn.CellTemplate>
                </DataGridTemplateColumn>
                <DataGridTextColumn Header="Dato1" Binding="{Binding Dato1}"/>
                <DataGridTextColumn Header="Dato2" Binding="{Binding Dato2}"/>
            </DataGrid.Columns>
        </DataGrid>
    </Grid>
</Window>
```

This class code:

```
/// <summary>
/// Interaction logic for MainWindow.xaml
/// </summary>
public partial class MainWindow : Window, INotifyPropertyChanged
{
    public ObservableCollection<Datos> Hmis { get; } = new ObservableCollection<Datos>();
    private System.Threading.Timer timer;

    public MainWindow()
    {
        InitializeComponent();
        timer = new System.Threading.Timer(ActualizacionUI_Tick, null, 1000, 1000);
    }

    public event PropertyChangedEventHandler? PropertyChanged;

    private void ActualizacionUI_Tick(object data)
    {
        Datos dato1 = new Datos() { Dato1 = "1", Dato2 = "2" };

        Datos dato2 = new Datos() { Dato1 = "1", Dato2 = "2" };

        Datos dato3 = new Datos() { Dato1 = "1", Dato2 = "2" };

        Dispatcher.Invoke(() =>
        {
            dato1.Estado = new SolidColorBrush((Color)ColorConverter.ConvertFromString("#FF5DC75D"));
            dato2.Estado = new SolidColorBrush((Color)ColorConverter.ConvertFromString("#FF5DC75D"));
            dato3.Estado = new SolidColorBrush((Color)ColorConverter.ConvertFromString("#FF5DC75D"));
            Hmis.Clear();
            Hmis.Add(dato1);
            Hmis.Add(dato2);
            Hmis.Add(dato3);
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs("Hmis"));
        });
    }
}
```

And this data class:

```
public class Datos : INotifyPropertyChanged
{
    public Brush Estado { get; set; }
    public string Dato1 { get; set; }
    public string Dato2 { get; set; }

    public event PropertyChangedEventHandler? PropertyChanged;
}
```

---
#### Feedback Bot on 10/20/2021, 11:07 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

**Comment by singhashish-wpf** [comment]
- Which version of .Net Core/Framework is being used?
- Please add sample app for your issue, so that we can check it.

**Comment by sneik15** [comment]
Probed in Framework versions 4.6.1/4.8 and net 5:
Example SLN:
[TremendoMemoryLeak.zip](https://github.com/dotnet/wpf/files/7426771/TremendoMemoryLeak.zip)
Only happens in visual studio 2022

**Comment by singhashish-wpf** [comment]
@sneik15 Have you also checked by using DependencyProperty for datos or providing implementation of INotifyPropertyChanged.

Please refer to https://stackoverflow.com/questions/18542940/can-bindings-create-memory-leaks-in-wpf/18543350#18543350

**Comment by sneik15** [comment]
@singhashish-wpf I have tested it and it still happens

**Comment by singhashish-wpf** [comment]
Great, could you please upload the zip for it also?
We do see that the memory usage increased for the app in diagnostics, however we also see new objects being created each time the callback is getting called; also, those objects are there in binding. So, we need to make sure this isn't an app specific issue.

**Comment by sneik15** [comment]
Both the list of the data(ObservableCollection) and the data(Datos) itself already implement INotifyPropertyChanged in the original zip that I have uploaded.

**Comment by singhashish-wpf** [comment]
> @singhashish-wpf Ashish Singh FTE I have tested it and it still happens

@sneik15 Please upload DependencyProperty version of the zip which you tested.

**Comment by sneik15** [comment]
@singhashish-wpf  I don't know how to correctly implement a DependencyProperty  version.

**Comment by brinko99** [comment]
I'm seeing an issue that I believe duplicates this. See attached for a very simple reproduction.

The leak appears to stem from making a selection on the `DataGrid` and relates to the debugger. The object bound to the selected row fails to be collected under VS2022 when debugging. Indeed, the behavior is different under VS2019 or when outside of the debugger. We see this with Core 3.1 and NET 6.

Reproduction using the attached:

VS2022:
1. Upon startup, two objects are added to an `ObservableCollection` bound to a DataGrid's `ItemsSource`
2. Select the row with ID:1
3. Press Clear (clears the collection)
4. Press Collect (forces GC)
5. Note in log that object ID:1 was not collected.
6. Press Create (adds two new items)
7. Select the row with ID:3
8. Press Clear
9. Press Collect
10. Note that again, only ID:2 was collected. ID:1 remains orphaned

![image](https://user-images.githubusercontent.com/1231765/152075318-ea95f6b6-f3de-403a-8f6d-1e59efae740b.png)

VS2019 or no debugger:
At step 10 per above, you'll see that ID:1 and ID:2 are collected
![image](https://user-images.githubusercontent.com/1231765/152075220-a2dc83ab-bfe0-491b-b7b3-bca2a2316271.png)

The diagnostic tools memory snapshot shows the following as the relevant Paths to Root for the orphaned object:
```
WpfDataGridLeak.Container
	ItemsControl+ItemInfo
		DataGrid
	ItemsControl+ItemInfo
		DataGridCellsPresenter
```


[WpfDataGridLeak.zip](https://github.com/dotnet/wpf/files/7982902/WpfDataGridLeak.zip)

**Reply by sneik15** [reply]
if you see the SNL i passed in the zip, the bug happens even without debugger(Only VS2022), so it will be another problem to add

**Comment by tomasfil** [comment]
Hello, I have the same issue. I originaly thought this was issue of ReactiveUI:
https://github.com/reactiveui/ReactiveUI/issues/3178

but it turns out it is issue with WPF somewhere.

I have built repository showcasing the issue (Branch without ReactiveUI and project ReactiveUI_DataGrid_MemoryLeak):
https://github.com/tomasfil/ReactiveUI_Tests/tree/WithoutReactiveUI/ReactiveUI_DataGrid_MemoryLeak

This issue exists for me on VS 2022, VS2019, no matter debugger or not, NET5, NET6, NET CORE 3.1

The demo logic is is MainWindow.xaml.cs
